File: 0044-wildcard-matching/0044-wildcard-matching.cpp
class Solution {
public:
    bool f(int i, int j, string &pattern, string &text, vector<vector<int>> &dp){
        if(i == 0 && j == 0) return true;
        if(i == 0 && j >0 ) return false;
        if(j == 0 && i > 0){
            for(int it = 1; it <=i; it++){
                if(pattern[it-1]!='*') return false;
            }
            return true;
        }
        if(dp[i][j] != -1) return dp[i][j];

        if(pattern[i-1] == text[j-1] || pattern[i-1] == '?'){
            return dp[i][j] = f(i-1, j-1, pattern, text, dp);
        }

        if(pattern[i-1] == '*'){
            return dp[i][j] =  f(i-1,j, pattern, text, dp) | f(i, j-1, pattern, text, dp);
        }

        return dp[i][j] = false;
    }

    bool isMatch(string s, string p) {
        int n = p.size();
        int m = s.size();
        vector<vector<int>> dp(n+1, vector<int>(m+1, -1));
        return f(n, m, p, s, dp);
    }
};

File: 0044-wildcard-matching/0044-wildcard-matching.py
class Solution:
    def findMatch(self, i, j, s, p, dp):
        if i < 0 and j < 0:
            return True
        if i < 0 and j >= 0:
            return False
        if i >= 0 and j < 0:
            for it in range(i+1):
                if p[it] != '*':
                    return False
            return True

        if dp[i][j] != -1:
            return dp[i][j]

        if p[i] == s[j] or p[i] == '?':
            dp[i][j] = self.findMatch(i-1, j-1, s, p, dp)
            return dp[i][j]
        
        if p[i] == '*':
            dp[i][j] = self.findMatch(i, j-1, s, p, dp) or self.findMatch(i-1, j, s, p, dp)
            return dp[i][j]

        dp[i][j] = False

        return dp[i][j]

    def isMatch(self, s: str, p: str) -> bool:
        n = len(p)
        m = len(s)

        dp = [[-1 for _ in range(m+1)] for _ in range(n+1)]

        return self.findMatch(n-1, m-1, s, p, dp)